### 一、数据的合并

数据合并主要包括下面两种操作：

* 轴向连接（concatenation），`pd.concat()`可以沿一个轴将多个DataFrame对象连接在一起，形成一个新的DataFrame对象

* 融合（merging），`pd.merge()`方法可以根据一个或多个键将不同DataFrame中的行连接起来

***

## concat

`concat()`函数可以将数据根据不同的轴作进行合并。我们先看一下`concat()`的常用参数：

```python
pd.concat(objs, axis=0, join='outer')
```

> **objs**: series，dataframe或者是panel构成的序列lsit 

> **axis**： 需要合并链接的轴，0是行，1是列，默认是0

> **join**：连接的方式 inner，或者outer，默认是outer（如果为inner，得到的是两表的交集，如果是outer，得到的是两表的并集。）

In [ ]:
表与表之间的数据处理：
交集/并集处理
将一个表的某一列，合并到另一张表

In [3]:
import pandas as pd
dict1={
    'A': ['A0', 'A1', 'A2', 'A3'],
    'B': ['B0', 'B1', 'B2', 'B3'],
    'C': ['C0', 'C1', 'C2', 'C3']}
df1=pd.DataFrame(dict1)
print(df1)

    A   B   C
0  A0  B0  C0
1  A1  B1  C1
2  A2  B2  C2
3  A3  B3  C3


In [4]:
dict2={
    'B': ['B0', 'B1', 'B2', 'B3'],
    'C': ['C0', 'C1', 'C2', 'C3'],
    'D': ['D0', 'D1', 'D2', 'D3']}
df2=pd.DataFrame(dict2)
print(df2)

    B   C   D
0  B0  C0  D0
1  B1  C1  D1
2  B2  C2  D2
3  B3  C3  D3


In [5]:
# 当`concat()`使用默认参数合并df1和df2时,`pd.concat([df1,df2],axis=0,join='outer')`。

In [20]:
pd.concat([df1,df2],axis=0,join='outer')  # axis参数为0时，列进行并集处理，纵向表拼接
                                          # axis参数为1时，行进行并集处理，横向表拼接，

# 通过上面的结果可以发现，当join='outer'，axis参数为0时，列进行并集处理，纵向表拼接，缺失值由NaN填充，并且会保留原有数据的行索引
# 如果两个表的index都没有实际含义，使用ignore_index参数，置true，重新整理一个新的index。

/Users/jiazhixiang/workspace/AI/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


A   B   C    D
0   A0  B0  C0  NaN
1   A1  B1  C1  NaN
2   A2  B2  C2  NaN
3   A3  B3  C3  NaN
0  NaN  B0  C0   D0
1  NaN  B1  C1   D1
2  NaN  B2  C2   D2
3  NaN  B3  C3   D3

In [ ]:
pip install --upgrade pandas #-i https://pypi.douban.com/simple
pip install --upgrade pandas -i https://pypi.douban.com/simple
# FutureWarning：排序是因为未串联的轴未对齐。 未来版本
# 的熊猫将默认更改为不排序。

# 要接受将来的行为，请传递'sort = False'。

# 要保留当前行为并消除警告，请传递“ sort = True”。

#    “”“启动IPython内核的入口点。
# ＃

In [16]:
# 如果两个表的index都没有实际含义，使用ignore_index参数，置true，重新整理一个新的index。
pd.concat([df1,df2],axis=0,join='outer',ignore_index=True)

/Users/jiazhixiang/workspace/AI/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


A   B   C    D
0   A0  B0  C0  NaN
1   A1  B1  C1  NaN
2   A2  B2  C2  NaN
3   A3  B3  C3  NaN
4  NaN  B0  C0   D0
5  NaN  B1  C1   D1
6  NaN  B2  C2   D2
7  NaN  B3  C3   D3

In [17]:
pd.concat([df1,df2],axis=1,join='outer')  # axis参数为1时，行进行并集处理，横向表拼接，

# 当join='outer'，axis参数为1时，行进行并集处理，横向表拼接，缺失值由NaN填充。
# 当concat()的join参数为inner时合并df1和df2时：

A   B   C   B   C   D
0  A0  B0  C0  B0  C0  D0
1  A1  B1  C1  B1  C1  D1
2  A2  B2  C2  B2  C2  D2
3  A3  B3  C3  B3  C3  D3

In [21]:
pd.concat([df1,df2],axis=0,join='inner')

B   C
0  B0  C0
1  B1  C1
2  B2  C2
3  B3  C3
0  B0  C0
1  B1  C1
2  B2  C2
3  B3  C3

## merge()

* 融合（merging），`pd.merge()`方法可以根据一个或多个键将不同DataFrame中的行连接起来

In [ ]:
merge()函数通过指定连接键拼接列数据，我们先看一下merge的常用参数：

merge(left, right, how='inner', on=None)
参数介绍
left和right：两个要合并的DataFrame
how：连接方式，有inner、left、right、outer，默认为inner
on：指的是用于连接的列索引名称，必须存在于左右两个DataFrame中，如果没有指定且其他参数也没有指定，则以两个DataFrame列名交集作为连接键

In [24]:
import pandas as pd
left = pd.DataFrame({'key':['a','b','b','d'],'data1':range(4)})
print(left)

  key  data1
0   a      0
1   b      1
2   b      2
3   d      3


In [25]:
right = pd.DataFrame({'key':['a','b','c'],'data2':range(3)})
print(right)

  key  data2
0   a      0
1   b      1
2   c      2


In [26]:
# 当merge()使用默认参数连接两个DataFrame时：

pd.merge(left, right)
# merge()默认做inner连接，并且使用两个DataFrame的列名交集（key）作为连接键，同样，最终连接的数据也是两个DataFramekey列数据的交集。

key  data1  data2
0   a      0      0
1   b      1      1
2   b      2      1

In [27]:
# 当两个DataFram使用做outer连接时：
pd.merge(left,right,on=['key'],how='outer')
# 当merge()做outer连接时最终连接的数据是两个DataFramekey列数据的并集，缺失的内容由NaN填充。

key  data1  data2
0   a    0.0    0.0
1   b    1.0    1.0
2   b    2.0    1.0
3   d    3.0    NaN
4   c    NaN    2.0

In [28]:
# 当两个DataFram使用left做连接时：
pd.merge(left,right,on=['key'],how='left')
# 当merge()做left连接时，最终连接的数据将以left数据的链接建为准,合并两个数据的列数据，缺失的内容由NaN填充

key  data1  data2
0   a      0    0.0
1   b      1    1.0
2   b      2    1.0
3   d      3    NaN

# 二、数据的筛选

文件路径：`../data/mouhu_users_2017.csv`

数据详情:id、关注的收藏夹数量、关注数量、关注者数量（粉丝数）、关注的问题数量、关注的话题数量、关注的专栏数量。

In [29]:
# 了解了数据的基本情况之后，第一个需求是将关注者超过100的用户数据获取出来。
import pandas as pd
df = pd.read_csv('../data/mouhu_users_2017.csv')
bools= df['关注者']>100
df1 = df[bools]
print(df1.shape)
print(df1)

(5501, 7)
                            _id  关注的收藏夹   关注   关注者  关注的问题  关注的话题  关注的专栏
79     587598f89f11daf90617fc18       8  111  3049   1216     45     17
101    587598f89f11daf90617fc44      61  123   205    670     24     18
112    587598f99f11daf90617fc5a       1  804   197    830     39     26
114    587598f99f11daf90617fc5e      44   62   201    314      8      3
121    587598f99f11daf90617fc6c      44  628  6597   1566    144     69
...                         ...     ...  ...   ...    ...    ...    ...
72694  5878399e9f11da09c8fe690e       3  172   219    466     68     11
72700  5878399e9f11da09c8fe691a      21   39   281     51     45      2
72713  5878399e9f11da09c8fe6934      49  635   248   1772    151    123
72726  5878399e9f11da09c8fe694e     447   62   728     14      1     21
72741  5878399e9f11da09c8fe696c    3546  390   142     25      4    126

[5501 rows x 7 columns]


In [30]:
# 我们已经准确获取到所有关注者超过100的用户数据，下面我们看一下代码的逻辑。

In [31]:
# 首先判断每个用户的关注者数量是否大于100，大于则会返回True，表示该行被标记为True，否则被标记为False。
# bools记录了每一行是否符合筛选条件，是一个Series对象，其中的值是bool类型。

In [33]:
df1 = df[bools]
df1

_id  关注的收藏夹   关注   关注者  关注的问题  关注的话题  关注的专栏
79     587598f89f11daf90617fc18       8  111  3049   1216     45     17
101    587598f89f11daf90617fc44      61  123   205    670     24     18
112    587598f99f11daf90617fc5a       1  804   197    830     39     26
114    587598f99f11daf90617fc5e      44   62   201    314      8      3
121    587598f99f11daf90617fc6c      44  628  6597   1566    144     69
...                         ...     ...  ...   ...    ...    ...    ...
72694  5878399e9f11da09c8fe690e       3  172   219    466     68     11
72700  5878399e9f11da09c8fe691a      21   39   281     51     45      2
72713  5878399e9f11da09c8fe6934      49  635   248   1772    151    123
72726  5878399e9f11da09c8fe694e     447   62   728     14      1     21
72741  5878399e9f11da09c8fe696c    3546  390   142     25      4    126

[5501 rows x 7 columns]

In [34]:
# 然后，根据**bools**每行的值来对df进行筛选，值为True，表示对应的行会留下，否则，则去除。

# 最后打印的df1数据就是关注者超过100的用户数据。这是pandas根据某列的值进行筛选的基本逻辑。

# 根据某列的值进行筛选的逻辑我们已经掌握了，如何进行多条件的联合筛选呢？

In [35]:
import pandas as pd
df = pd.read_csv('../data/mouhu_users_2017.csv')
bool1= df['关注者']>300
bool2= df['关注']>100
df2 = df[bool1 & bool2]
df2.head()

_id  关注的收藏夹   关注    关注者  关注的问题  关注的话题  关注的专栏
79   587598f89f11daf90617fc18       8  111   3049   1216     45     17
121  587598f99f11daf90617fc6c      44  628   6597   1566    144     69
155  587598f99f11daf90617fcb0     194  250   1103     10      1     19
228  587598f99f11daf90617fd42      48  224   1450    360    128     20
261  587598f99f11daf90617fd84    1125  842  24848    108     33     27

In [ ]:
# 上面的这段代码里，我们通过了2个限制条件df['关注者']>300和 df['关注']>100，分别得到 bool1和bool2这2个Series。

# 在我们的需求中，需要的数据是同时满足两个条件，所以我们使用逻辑**与**运算连接两个值，最后获取同时满足两个条件的值。

# 数据的排序

In [ ]:
# 在数据获取过程中，数据的排序也是我们经常需要处理的问题
# 可以使用`sort_index()`、`sort_values()`两个方法对数据进行排序，并且这两个方法Series和DataFrame都支持。
# DataFrame的`sort_index( )`方法是按照行索引进行排序，`sort_values()`可以指定具体列进行排序。

In [ ]:
获取了关注者数据前十的用户数据

In [37]:
import pandas as pd
df = pd.read_csv('../data/mouhu_users_2017.csv')
df.sort_values(by='关注者',ascending=False,inplace=True)
print(df.head(10))

                            _id  关注的收藏夹    关注     关注者  关注的问题  关注的话题  关注的专栏
23313  5876ee159f11da03c9726155    3181  1053  585787     93     33     26
14461  5876ede89f11da03c9721c33    1041   254  356246     92      3     23
22048  5876ee0c9f11da03c9725773      41    72  351414    336     47     16
52464  5876ece59f11da03afbc71b3    1652   794  308553     23      1    112
42135  5876eb5a9f11da0399d461b7    1182   532  268353     23      1     54
63579  5878396e9f11da09c8fe21e6   10819  1412  198199     90      1    517
21740  5876ee0b9f11da03c972550b     991   479  195175     46      3     18
56721  5876ecfa9f11da03afbc92f1     164   550  164254      0      1     19
22784  5876ee129f11da03c9725d33    2011   579  148457     49      6     44
29577  587838f59f11da09bd427945       8   303  137087    477     19     15


我们成功的获取了关注者数据前十的用户数据，按照惯例我们一起分析一下代码。

`by`:决定了是按数据中的哪一列进行排序，将需要按照某列排序的列名赋值给by即可

`ascending=False`:将数据按照从大到小的顺序排列

`inplace=True`:用来控制是否直接对原始数据进行修改

根据排序后的结果，我们发现最高的关注者数量已经达到了585787，并远远的看着第二名的356246，确有大V的潜质啊。
